### PR TITLE
Update RNW CLI to preserve and re-use `init-windows` arguments

### LINKED ISSUE #13174: Autolinking for C++ projects only generates AutolinkedNativeModules.g.cpp, not the header file
### Problem Description

The header file is necessary, and while it doesn't really change when autolinking is run, autolinking should still (re)create the file, as it may have been deleted by the end user (or missing for another reason) and we claim in the file that it gets generated by the autolink command, when it clearly isn't.

### Steps To Reproduce

1. Create a new C++ RNW project.
2. Delete AutolinkedNativeModules.g.h
3. Run autolink-windows or run-windows

### Expected Results

The file gets recreated.

### CLI version

npx react-native -v

### Environment

```markdown
npx react-native info
```

### Target Platform Version

None

### Target Device(s)

_No response_

### Visual Studio Version

None

### Build Configuration

None

### Snack, code example, screenshot, or link to a repository

_No response_

### LINKED ISSUE #13926: Autolinking should use the correct template as source files for autolinking codegen
### Problem Description

Right now, autolinking always uses the template files from the old templates to generate the right includes. This precludes us from differentiating what those files look like in fabric.

Autolinking should load the correct template files for fabric instead of using the paper ones.

### Steps To Reproduce

Run autolinking on fabric app project

### Expected Results

_No response_

### CLI version

npx @react-native-community/cli -v

### Environment

```markdown
npx @react-native-community/cli info
```

### Community Modules

_No response_

### Target Platform Version

None

### Target Device(s)

_No response_

### Visual Studio Version

None

### Build Configuration

None

### Snack, code example, screenshot, or link to a repository

_No response_

### LINKED ISSUE #13927: init-windows should stamp the name of the template used into new projects
### Summary

Expo uses a system of stamping a platform config object into the RN app's app.json file to use later with its commands. We should do something similar to stamp the template being used, something like:

```json
"windows": {
    "template": "cpp-app"
}
```

### Motivation

Having this information in the app.json would make it easier for our commands (like autolinking) to use that information.

### Basic Example

_No response_

### Open Questions

_No response_

### LINKED ISSUE #13937: `init-windows` should identify the correct default template based on the existing code
If the project appears to be using an existing template (perhaps detecting it because of #13927 ) then use that template again.

If no project is identified, try to determine where to use an app or library template.

The final fallback (if we have nothing to go on) do the default template.

### PR DESCRIPTION
**Pull request by jonthysell**
## Description

This PR updates the RNW CLI to preserve and re-use the arguments passed to `init-windows` when the project was created.

### Type of Change
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

### Why
The RNW CLI should be smart enough to change its command's behaviors based on the template used by the project.

Closes #13927 
Closes #13926 
Closes #13937 
Closes #13174 

### What

As per #13927:
1. The `init-windows` command will now stamp the `template`, `name`, and `namespace` arguments to the project's `package.json` file as follows:
    ```json
    {
      "react-native-windows": {
        "init-windows": {
          "name": "testcli",
          "namespace": "testcli",
          "template": "cpp-app",
        },
      },
    }
    ```
2. The `config` command (and therefore all other CLI commands) will now contain the contents of this new `react-native-windows` section in a new `rnwConfig` property.

Then, as per #13937:
1. The `init-windows` command will now use the previously saved arguments if they're available (and not otherwise overwritten by passed-in arguments)
2. If the template arg is missing (as it will be for all previously created projects) the `config` command will try to determine which template was used and inject it

Additionally, as per #13926:
1. The `autolink-windows` command will now use the template information to use the correct files when autolinking

Bonus, as per #13174:
1. The `autolink-windows` command will now include both the CPP and H file (the H file wasn't being regenerated before)

## Screenshots
N/A

## Testing
Verified new projects contain the new information.

## Changelog
Should this change be included in the release notes: _yes_

Update RNW CLI to preserve and re-use `init-windows` arguments
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/react-native-windows/pull/13988)

### CODE REVIEW

**Review by TatianaKapos — APPROVED** [review]
LGTM!

File: packages/@react-native-windows/cli/src/commands/autolinkWindows/autolinkWindows.ts [hunk 1]
```diff
@@ -340,27 +377,21 @@ export class AutoLinkWindows {
     if (cppPackageProviders === '') {
       // There are no windows dependencies, this would result in warning. C4100: 'packageProviders': unreferenced formal parameter.
       // therefore add a usage.
-      cppPackageProviders = '\n    UNREFERENCED_PARAMETER(packageProviders);'; // CODESYNC: vnext\local-cli\generator-windows\index.js
+      cppPackageProviders = '\n    UNREFERENCED_PARAMETER(packageProviders);'; // CODESYNC: @react-native-windpows\cli\src\generator-windows\index.ts
```
Comment (danielayala94) [inline packages/@react-native-windows/cli/src/commands/autolinkWindows/autolinkWindows.ts:380]: nit: typo at `windpows`.

File: packages/@react-native-windows/cli/src/commands/autolinkWindows/autolinkWindows.ts [hunk 2]
```diff
@@ -284,23 +278,51 @@ export class AutoLinkWindows {
...
+        case '.h':
+          templateRoot = path.join(templateRoot, 'cpp-app', 'src');
+          break;
+        case '.cs':
+          templateRoot = path.join(templateRoot, 'cs-app', 'src');
+          break;
+        case '.props':
+        case '.targets':
+        default:
+          templateRoot = path.join(templateRoot, 'shared-app', 'src');
```
Comment (danielayala94) [inline packages/@react-native-windows/cli/src/commands/autolinkWindows/autolinkWindows.ts:298]: Just curious, what is a `shared-app`?

File: packages/@react-native-windows/cli/src/commands/autolinkWindows/autolinkWindows.ts (hunk 1 above)
Comment (danielayala94) [reply packages/@react-native-windows/cli/src/commands/autolinkWindows/autolinkWindows.ts:380]: Same with other usages.

File: vnext/templates/templateUtils.js [hunk 3]
```diff
@@ -110,7 +110,12 @@ async function runNpmInstall(config = {}, options = {}) {
   }
 }
 
-async function updateProjectPackageJson(config = {}, options = {}, props = {}) {
+async function updateProjectPackageJson(
+  config = {},
+  options = {},
+  props = {},
+  saveOptions = true,
```
Comment (danielayala94) [inline vnext/templates/templateUtils.js:117]: Is there a case where `saveOptions` is false?

Comment (jonthysell) [reply vnext/templates/templateUtils.js:117]: Not yet, but it's a new behavior that we may want to disable for some templates, I'm not sure yet.

File: packages/@react-native-windows/cli/src/commands/autolinkWindows/autolinkWindows.ts (hunk 1 above)
Comment (jonthysell) [reply packages/@react-native-windows/cli/src/commands/autolinkWindows/autolinkWindows.ts:380]: In the future, you can use the suggestion button in your comment to suggesta  typo fix that I can then just accept, as so:

```suggestion
      cppPackageProviders = '\n    UNREFERENCED_PARAMETER(packageProviders);'; // CODESYNC: @react-native-windows\cli\src\generator-windows\index.ts
```

File: packages/@react-native-windows/cli/src/generator-windows/index.ts [hunk 4]
```diff
@@ -227,7 +227,7 @@ export async function copyProjectTemplateAndReplace(
     autolinkCsReactPackageProviders: '',
     autolinkCppIncludes: '',
     autolinkCppPackageProviders:
-      '\n    UNREFERENCED_PARAMETER(packageProviders);', // CODESYNC: vnext\local-cli\runWindows\utils\autolink.js
+      '\n    UNREFERENCED_PARAMETER(packageProviders);', // CODESYNC: @react-native-windpows\cli\src\commands\autolinkWindows\autolinkWindows.ts
```
Comment (jonthysell) [inline packages/@react-native-windows/cli/src/generator-windows/index.ts:230]:
```suggestion
      '\n    UNREFERENCED_PARAMETER(packageProviders);', // CODESYNC: @react-native-windows\cli\src\commands\autolinkWindows\autolinkWindows.ts
```

File: vnext/templates/cpp-app/template.config.js [hunk 5]
```diff
@@ -75,6 +75,13 @@ async function getFileMappings(config = {}, options = {}) {
     addReactNativePublicAdoFeed: true || isCanary, // Temporary true for all new projects until code-signing is restored, see issue #14030
 
     cppNugetPackages,
+
+    // autolinking template variables
+    autolinkPropertiesForProps: '',
+    autolinkProjectReferencesForTargets: '',
+    autolinkCppIncludes: '',
+    autolinkCppPackageProviders:
+      '\n    UNREFERENCED_PARAMETER(packageProviders);', // CODESYNC: @react-native-windpows\cli\src\commands\autolinkWindows\autolinkWindows.ts
```
Comment (jonthysell) [inline vnext/templates/cpp-app/template.config.js:84]:
```suggestion
      '\n    UNREFERENCED_PARAMETER(packageProviders);', // CODESYNC: @react-native-windows\cli\src\commands\autolinkWindows\autolinkWindows.ts
```

File: packages/@react-native-windows/cli/src/commands/autolinkWindows/autolinkWindows.ts (hunk 2 above)
Comment (jonthysell) [reply packages/@react-native-windows/cli/src/commands/autolinkWindows/autolinkWindows.ts:298]: The old paper templates were grouped into sets of folders based on type (shared by all apps, shared by all libs, project vs source, etc) so they could be re-used by multiple templates. However this meant a bunch of custom logic in the generator to decide which files to use in which situations.

With `init-windows`, the decision was to simplify the matrix of options into named templates, where you could open a single folder with its literal name and see exactly which files it was going to copy in a new project.

**Pull request merged**